### PR TITLE
Dynamic Configuration

### PR DESCRIPTION
**Pull request by Integralist**
**Problem**: there are multiple hardcoded global configuration values that require a new CLI release to be updated.
**Solution**: pull configuration from an API endpoint.
**Notes**: 
The long term intention would be for the configuration to live under `api.fastly.com` but currently that means the response should be in JSON format, which the configuration is _not_ (it's toml, which would mean a substantial change to the code in this PR and would in general be considered a negative impact on the implementation). Subsequently we've placed the config under `developer.fastly.com` and defined the path as `/api/internal/cli-config` to signal that there is no backwards compatibility guarantees.

To save people from scanning the code in this PR, below is:

- Example toml configuration.
- Example VCL service to serve toml configuration.
- Example golang structures to marshal toml configuration.

---

Here is the example toml configuration:

```toml
[fastly]
api_endpoint = "https://api.fastly.com"

[cli]
remote_config = "https://developer.fastly.com/api/internal/cli-config"
ttl = "5m"
last_checked = "2021-01-21T16:07:20Z"

[[user]]
[[user.profile]]
name = "work"
email = "integralist@fastly.com"
token = "..."
[[user.profile]]
name = "personal"
email = "hello@integralist.co.uk"
token = "..."

[language]
  [language.rust]
  toolchain_version = "1.49.0"
  wasm_wasi_target = "wasm32-wasi"
  fastly_sys_constraint = ">= 0.3.7 <= 0.6.0"

[starter-kits]
[[starter-kits.assemblyscript]]
  name = "Default"
  path = "https://github.com/fastly/compute-starter-kit-assemblyscript-default"
  tag = "v0.2.0"
[[starter-kits.rust]]
  name = "Default"
  path = "https://github.com/fastly/compute-starter-kit-rust-default.git"
  branch = "0.6.0"
[[starter-kits.rust]]
  name = "Beacon"
  path = "https://github.com/fastly/compute-starter-kit-rust-beacon-termination.git"
[[starter-kits.rust]]
  name = "Static"
  path = "https://github.com/fastly/compute-starter-kit-rust-static-content.git"
```

Here is the simple api backend code (this will be replaced by an internal fastly implementation):
```vcl
sub vcl_recv {
  #FASTLY recv
  
  if (req.url.path ~ "^/cli/config$") {
    error 600 "Synthetic TOML";
  }
  error 404 "Not Found";
}

sub vcl_error {
  #FASTLY error
  
  if (obj.status == 404) {
    set obj.status = 404;
    set obj.response = "Not Found";
    set obj.http.Content-Type = "text/html; charset=UTF-8";
    synthetic {"
      	<!doctype html>
        <html lang="en">
          <head>
            <meta charset="utf-8">
            <title>Whoops!</title>
          </head>
          <body>
            <h1>FAIL</h1>
          </body>
        </html>
    "};
    
    return(deliver);
  }
  
  if (obj.status == 600) {
    set obj.status = 200;
    set obj.response = "OK";
    set obj.http.Content-Type = "application/toml; charset=utf8";
    
    # we omit sections (e.g. [user]) and keys (e.g. last_checked)
    # as we either shouldn't be setting those at the API layer
    # e.g. we don't know anything about a 'user'
    # or we want the CLI to handle generating the value
    # e.g. the last_checked value we determine within the CLI
    #
    synthetic {"[fastly]
api_endpoint = "https://api.fastly.com"

[cli]
remote_config = "http://developer.fastly.net/cli/config"
ttl = "5m"

[language]
  [language.rust]
  toolchain_version = "1.49.0"
  wasm_wasi_target = "wasm32-wasi"
  fastly_sys_constraint = ">= 0.3.7 <= 0.6.0"
  
[starter-kits]
[[starter-kits.assemblyscript]]
  name = "Default"
  path = "https://github.com/fastly/compute-starter-kit-assemblyscript-default"
  tag = "v0.2.0"
[[starter-kits.rust]]
  name = "Default"
  path = "https://github.com/fastly/compute-starter-kit-rust-default.git"
  branch = "0.6.0"
[[starter-kits.rust]]
  name = "Beacon"
  path = "https://github.com/fastly/compute-starter-kit-rust-beacon-termination.git"
[[starter-kits.rust]]
  name = "Static"
  path = "https://github.com/fastly/compute-starter-kit-rust-static-content.git""};
    
    return(deliver);
  }
}
```

Here is the go structure the toml is mapped to:

```go
// File represents our dynamic application toml configuration.
type File struct {
	Fastly      ConfigFastly              `toml:"fastly"`
	CLI         ConfigCLI                 `toml:"cli"`
	User        ConfigUser                `toml:"user"`
	Language    ConfigLanguage            `toml:"language"`
	StarterKits ConfigStarterKitLanguages `toml:"starter-kits"`
}

// ConfigFastly represents fastly specific configuration.
type ConfigFastly struct {
	APIEndpoint string `toml:"api_endpoint"`
}

// ConfigCLI represents CLI specific configuration.
type ConfigCLI struct {
	RemoteConfig string `toml:"remote_config"`
	TTL          string `toml:"ttl"`
	LastChecked  string `toml:"last_checked"`
}

// ConfigUser represents user specific configuration.
type ConfigUser struct {
	Token string `toml:"token"`
	Email string `toml:"email"`
}

// ConfigLanguage represents C@E language specific configuration.
type ConfigLanguage struct {
	Rust ConfigRust `toml:"rust"`
}

// ConfigLanguage represents Rust C@E language specific configuration.
type ConfigRust struct {
	// ToolchainVersion is the `rustup` toolchain string for the compiler that we
	// support
	ToolchainVersion string `toml:"toolchain_version"`

	// WasmWasiTarget is the Rust compilation target for Wasi capable Wasm.
	WasmWasiTarget string `toml:"wasm_wasi_target"`

	// FastlySysConstraint is a free-form semver constraint for the internal Rust
	// ABI version that should be supported.
	FastlySysConstraint string `toml:"fastly_sys_constraint"`
}

// ConfigStarterKitLanguages represents language specific starter kits.
type ConfigStarterKitLanguages struct {
	AssemblyScript []ConfigStarterKit `toml:"assemblyscript"`
	Rust           []ConfigStarterKit `toml:"rust"`
}

// ConfigStarterKit represents starter kit specific configuration.
type ConfigStarterKit struct {
	Name     string `toml:"name"`
	Path     string `toml:"path"`
	Tag      string `toml:"tag"`
	Branch   string `toml:"branch"`
}
```

### Legacy Configuration Format

If you have a configuration file using the legacy format (see below), then we will automatically migrate your token and email.

```toml
token = "..."
email = "integralist@fastly.com"
endpoint = "https://api.fastly.com"
last_version_check = "2021-02-11T11:48:52Z"
```

We'll extract the `token` and `email` fields and place them in the correct place within the new format:

```
...

[user]
  token = "..."
  email = "integralist@fastly.com"

...
```

### CODE REVIEW

File: pkg/compute/assemblyscript.go
Comment (Integralist) [inline pkg/compute/assemblyscript.go:56]: I wanted to use an existing function `MakeDirectoryIfNotExists` but I couldn't because of a cyclic import issue, so I needed to move those functions into a separate module (as `common` appeared to be containing too much miscellaneous behaviour).

File: pkg/config/data.go
Comment (Integralist) [inline pkg/config/data.go:128]: Not sure if this is useful, but could be interesting if at some point we wanted to dynamically change the backend serving up the configuration data.

File: pkg/config/data.go
Comment (Integralist) [inline pkg/config/data.go:159]: Not sure if this will be OK for non-unix systems 🤷🏻

File: pkg/config/data.go
Comment (phamann) [reply pkg/config/data.go:128]: Isn't this what the `file.Endpoint` field on line 124 above is already for?

File: pkg/config/data.go
Comment (phamann) [reply pkg/config/data.go:159]: I believe you want [`filepath.Base()`](https://golang.org/pkg/path/filepath/#Base) which should be safe cross-platform

File: pkg/config/data.go
Comment (Integralist) [reply pkg/config/data.go:128]: oh really!? lol ok then 👍🏻 🙂

File: pkg/config/data.go
Comment (Integralist) [reply pkg/config/data.go:159]: Ah nice!

Comment (Integralist) [reply pkg/config/data.go:159]: @phamann Actually I don't think that is what we want as it gives back `config.toml` instead of the path leading to that section.

Comment (Integralist) [reply pkg/config/data.go:159]: Ah there we go `filepath.Dir`

File: cmd/fastly/main.go
Comment (Integralist) [inline cmd/fastly/main.go:19]: This would obviously be an `api.fastly.com` endpoint.

File: cmd/fastly/main.go
Comment (Integralist) [inline cmd/fastly/main.go:47]: As noted in my design doc... 

**Should we be informing users of the existence of this file?** 
Currently I am displaying messages to the user to inform them when the local configuration file is missing and whether it’s stale, and whether we’ve been able to update it successfully. This feels overly noisy and not necessarily of value to a user. It might be better to hide this output behind the `-v`, `-verbose` flag (or `-d`, `-debug` †)

> † as noted by https://clig.dev/#arguments-and-flags `-v` can be confused with `--version`.

**Comment by Integralist** [comment]
👋🏻 @acfoltzer here is the CLI 'dynamic config' PR. If you review the [last commit](https://github.com/fastly/cli/pull/187/commits/2b284171650e3f22302eb0780527a5920c75b652) you'll see the changes I've implemented to use a value pulled from the remote configuration object (specifically the file `pkg/compute/rust.go`).

File: pkg/compute/rust.go
Comment (acfoltzer) [inline pkg/compute/rust.go:222]: I believe this should be `semver.NewConstraint` in order to parse the kinds of ranges we'll want to support. For example, we want to make sure that the fastly-sys in a package is `>= 0.3.0 && < 0.5.0`, not that it equals a precise version.

Comment (Integralist) [reply pkg/compute/rust.go:222]: Aha, ok gotcha. I'll fix that up 👍🏻

**Comment by Integralist** [comment]
## TODO

- `semver.NewConstraint` instead of `semver.NewVersion`.
- don't include hardcoded failover, just error.
- implement new starter-kit toml syntax.

File: pkg/compute/rust.go
Comment (Integralist) [reply pkg/compute/rust.go:222]: Actually I just realize that we do create a constraint on a line just following this section. But it looks like we need an explicit min and max version for fastly-sys rather than constraining by a single value to the minor version.

File: pkg/compute/rust.go
Comment (acfoltzer) [inline pkg/compute/rust.go:222]: Can we just have this be a free-form constraint? If we ever have to yank a version of the crate we'd want this to be a disjoint range, not just a min-max.

Comment (Integralist) [reply pkg/compute/rust.go]: No problem. I've updated this 👍🏻

Comment (acfoltzer) [reply pkg/compute/rust.go]: Thank you! :tada:

File: pkg/compute/compute_integration_test.go
Comment (acfoltzer) [inline pkg/compute/compute_integration_test.go:647]: `fastly 0.3.3` is where `fastly-sys` was introduced, but the earlier `0.3.x` do not have it.

**Review by acfoltzer — COMMENTED** [review]
Approved from the Rust semantics perspective. Thank you so much!

I am not a Go programmer so I will leave it to someone else to give the genuine :heavy_check_mark:

**Review by kailan — CHANGES_REQUESTED** [review]
This looks awesome! Only a couple of comments on the transition from the old format to the new:

- When users have a config in the old format, the relevant values should be migrated over to the new format. When I ran this branch, I had to re-run `fastly configure` to set my token again.
- When users have a config in the old format, this logic currently seems to skip downloading the new config and instead irreversibly overwrites the config with empty data. (https://github.com/fastly/cli/pull/187/files#diff-3bb21ba08d70ab5924cf7e23ea51e7e68f73f0b164a0b23f1ec980b70a323104R55)

File: cmd/fastly/main.go
Comment (kailan) [inline cmd/fastly/main.go:81]: This could be misleading – "local application" makes it sound like it could be an issue with their C@E project. I would maybe explain that this is the "CLI configuration".

Also rather than out of date... "the Fastly CLI configuration has recently changed, we'll refresh your local copy"? Sounds less like a problem.

File: cmd/fastly/main.go
Comment (triblondon) [inline cmd/fastly/main.go:82]: 'local application' is likely to be mis-read as the customer's package, not the fastly CLI itself.

```suggestion
Compatibility and versioning information for the Fastly CLI is being updated in the background.  The updated data will be used next time you execute a fastly command.
```

File: pkg/config/data.go
Comment (triblondon) [inline pkg/config/data.go:51]:
```suggestion
	UpdateSuccessful = "Successfully updated platform compatibility and versioning information."
```

File: cmd/fastly/main.go
Comment (triblondon) [reply cmd/fastly/main.go:81]: Snap!  I had the same comment - though we can't say for sure that it's changed, only that the current version is out of date.  I made a suggestion in my review.

File: pkg/config/data.go
Comment (phamann) [inline pkg/config/data.go:161]:
```suggestion
// ConfigRust represents Rust C@E language specific configuration.
```

File: pkg/config/data.go
Comment (phamann) [inline pkg/config/data.go:168]: Does this need to be Dynamic configuration? I don't ever see this changing. cc @acfoltzer

File: pkg/config/data.go
Comment (phamann) [inline pkg/config/data.go:192]: 3 seconds seems very high? Also can we move this timeout to a named constant at the top of the file?

File: cmd/fastly/main.go
Comment (phamann) [inline cmd/fastly/main.go:47]: I don't think `-verbose` is valid?

```suggestion
		if seg == "-v" || seg == "--verbose" {
```

File: cmd/fastly/main.go
Comment (phamann) [inline cmd/fastly/main.go:81]: General quesiton which applies to a couple of output messages in this PR: whats the reasoning for _"Compatibility and versioning information"_ versus simply _"Configuration"_?

**Review by phamann — COMMENTED** [review]
Looking good, so happy to see how this has evolved! Couple of minor nits and questions.

File: cmd/fastly/main.go
Comment (kailan) [reply cmd/fastly/main.go:81]: Discussion [here](https://github.com/fastly/cli/pull/187#discussion_r574444593) and [here](https://github.com/fastly/cli/pull/187#discussion_r574486737).

Comment (phamann) [reply cmd/fastly/main.go:81]: That discussion is about the use of 'local application' - which I agree with. However, @triblondon's suggested edit of "compatibility and versioning information" seems strange to me and might not always be true in the future, where a generic "configuration" will.

Comment (triblondon) [reply cmd/fastly/main.go:81]: I don't love what I suggested here.  It was to solve the 'local application' problem, but 'configuration' is too broad imho.  I think we need to give the user a sense of what this data is for.

File: pkg/config/data.go
Comment (acfoltzer) [reply pkg/config/data.go:168]: It changed in the past from `wasm32-unknown-wasi` to `wasm32-wasi`, and while I doubt it'll change again we might as well be prepared

File: cmd/fastly/main.go
Comment (Integralist) [reply cmd/fastly/main.go:81]: I'm OK with this for now.

File: pkg/compute/compute_integration_test.go
Comment (Integralist) [inline pkg/compute/compute_integration_test.go:647]:
```suggestion
			wantError:            "fastly-sys crate not found", // fastly 0.3.3 is where fastly-sys was introduced
```

File: pkg/config/data.go
Comment (Integralist) [reply pkg/config/data.go]: @phamann considering we expect the response to be a lot quicker, is there any real danger in setting it to three seconds? In the case of a freak scenario where the response is slow I'd rather give the client a better chance at acquiring the file rather than setting it too low.

**Comment by Integralist** [comment]
@phamann @kailan I've addressed both your comments. 

If you could please take a look at the last 6 commits, re-review and approve if you're happy. Thanks!

**Review by kailan — APPROVED** [review]
Looks great! This is going to be so useful.

File: cmd/fastly/main.go
Comment (phamann) [inline cmd/fastly/main.go:60]: Just noting that I think this is the first time we've used a third person tone of voice "we" in the CLI output. And thus inconsistent with other messages. If thats our preferred tone going forward we should consider starting a document somewhere (internally or in this repo) with best practices for writing output messaging for the product. I personally found the use of "we" here strange 😉

File: cmd/fastly/main.go
Comment (phamann) [inline cmd/fastly/main.go:74]: Noting an outstanding `todo`, should we add a proper remediation step to this PR?

File: cmd/fastly/main.go
Comment (phamann) [reply cmd/fastly/main.go:88]: Still find this weird 😬

File: cmd/fastly/main.go
Comment (phamann) [inline cmd/fastly/main.go:100]: Another outstanding TODO?

**Review by phamann — APPROVED** [review]
LGTM - A couple of outstanding TODO items and some nits about use of language but none are blocking 😄

File: cmd/fastly/main.go
Comment (Integralist) [reply cmd/fastly/main.go:60]: So something like this instead?

```
Found local configuration file...
File is being upgraded.
```

**Comment by Integralist** [comment]
Tested this again locally...

```bash
go run cmd/fastly/main.go version -v
```

Using a legacy configuration file:

```toml
# /Users/integralist/Library/Application Support/fastly/config.toml

token = "abcdefg"
email = "integralist@fastly.com"
endpoint = "https://api.fastly.com"
last_version_check = "2021-02-11T11:48:52Z"
```

This caused the following output:

```
Found your local configuration file (required to use the CLI) was using a legacy format.
File is being upgraded now.

Fastly API token provided via config file
Fastly API endpoint: https://api.fastly.com
Fastly CLI version v0.0.0-unknown (unknown)
Built with go version unknown
```

This updated the legacy toml configuration to:

```toml
[cli]
  last_checked = "2021-03-15T16:57:50Z"
  remote_config = "https://developer.fastly.com/api/internal/cli-config"
  ttl = "5m"

[fastly]
  api_endpoint = "https://api.fastly.com"

[language]

  [language.rust]
    fastly_sys_constraint = ">= 0.3.7 <= 0.6.0"
    toolchain_version = "1.49.0"
    wasm_wasi_target = "wasm32-wasi"

[legacy]
  email = "integralist@fastly.com"
  token = "abcdefg"

[starter-kits]

  [[starter-kits.assemblyscript]]
    branch = ""
    name = "Default"
    path = "https://github.com/fastly/compute-starter-kit-assemblyscript-default"
    tag = "v0.2.0"

  [[starter-kits.rust]]
    branch = "0.6.0"
    name = "Default"
    path = "https://github.com/fastly/compute-starter-kit-rust-default.git"
    tag = ""

  [[starter-kits.rust]]
    branch = ""
    name = "Beacon"
    path = "https://github.com/fastly/compute-starter-kit-rust-beacon-termination.git"
    tag = ""

  [[starter-kits.rust]]
    branch = ""
    name = "Static"
    path = "https://github.com/fastly/compute-starter-kit-rust-static-content.git"
    tag = ""

[user]
  email = "integralist@fastly.com"
  token = "abcdefg"
```

**Pull request merged**
